scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -0,0 +1,158 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.mesos.optimizer
+  (:require [cheshire.core :as cheshire]
+            [chime :refer [chime-at chime-ch]]
+            [clj-http.client :as http]
+            [clj-time.core :as t]
+            [clj-time.coerce :as tc]
+            [clj-time.periodic :as periodic]
+            [clojure.core.async :as async]
+            [clojure.tools.logging :as log]
+            [cook.util :refer [lazy-load-var]]
+            [cook.mesos.util :as util]
+            [com.rpl.specter :as sp]
+            [datomic.api :as d :refer (q)]
+            [schema.core :as s]))
+(def PosNum
+  (s/both s/Num (s/pred pos? 'pos?)))
+
+(def PosInt
+  (s/both s/Int (s/pred pos? 'pos?)))
+
+(def NonNegInt
+  (s/both s/Int (s/pred #(<= 0 %) 'pos?)))
+
+(def TimePeriodMs NonNegInt)
+
+;; Elements missing:
+;;   1. Attributes (chip set, az, ..) -- have a combinitorial problem here
+;;   2. Price
+;;   3. Prob preemption in 10 minutes increments (or something like that
+(def HostInfo
+  {:count NonNegInt
+   :instance-type s/Str
+   :cpus PosNum
+   :mem PosNum
+   (s/optional-key :gpus) PosNum
```

Comment:
Should `cpus`, `gpus` and `mem` be placed under a `:resources` entry?

Reply:
I prefer this map being flat since it is so simple. Do you feel strongly about them being under a resources key?